File: src/lexer.rs
use crate::token::{Token, TokenTag};
use std::collections::VecDeque;

#[derive(Debug)]
pub struct Lexer<'a> {
    source: &'a str,
    index: usize,
    tokens: VecDeque<Token<'a>>,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            source,
            index: 0,
            tokens: VecDeque::new(),
        }
    }

    pub fn peek(&self) -> Option<&Token<'a>> {
        self.tokens.front()
    }

    pub fn next(&mut self) -> Option<Token<'a>> {
        self.tokens.pop_front()
    }

    pub fn process(&mut self) {
        while self.peek_char().is_some() {
            self.clear_whitespace();
            self.tokenize_next();
        }
    }

    fn tokenize_next(&mut self) {
        assert!(self.peek_char().is_some());
        let c = self.peek_char().unwrap();
        if c.is_numeric() {
            self.tokenize_number();
        } else if c.is_alphabetic() || c == '_' {
            self.tokenize_identifier();
        } else {
            self.tokenize_special();
        }
    }

    fn tokenize_number(&mut self) {
        assert!(self.peek_char().is_some());
        let (start, end) = self.tokenize_while(|c| c.is_numeric());
        self.tokens.push_back(Token {
            tag: TokenTag::Number,
            raw: &self.source[start..end],
        })
    }

    fn tokenize_identifier(&mut self) {
        assert!(self.peek_char().is_some());
        let (start, end) = self.tokenize_while(|c| c.is_alphabetic() || c == '_');
        self.tokens.push_back(Token {
            tag: TokenTag::Number,
            raw: &self.source[start..end],
        });
    }

    fn tokenize_special(&mut self) {
        assert!(self.peek_char().is_some());
        let start = self.index;
        let first = self.eat_char().unwrap();
        let tag = match first {
            '(' => TokenTag::LParen,
            ')' => TokenTag::RParen,
            '{' => TokenTag::LCurly,
            '}' => TokenTag::RCurly,
            _ => panic!("Unrecognized character: '{first}'"),
        };
        let end = self.index;
        self.tokens.push_back(Token {
            tag,
            raw: &self.source[start..end],
        })
    }

    fn clear_whitespace(&mut self) {
        self.tokenize_while(|c| c.is_whitespace());
    }

    fn tokenize_while(&mut self, func: impl Fn(char) -> bool) -> (usize, usize) {
        let start = self.index;
        while let Some(c) = self.peek_char() {
            if !func(c) {
                break;
            }
            self.eat_char();
        }
        let end = self.index;
        (start, end)
    }

    fn peek_char(&self) -> Option<char> {
        self.source.chars().nth(self.index)
    }

    fn eat_char(&mut self) -> Option<char> {
        let c = self.source.chars().nth(self.index);
        self.index += 1;
        c
    }
}

File: src/lexer.zig
const std = @import("std");

pub const TokenTag = enum {
    identifier,
    number,
    l_paren,
    r_paren,
    l_curly,
    r_curly,
    comma,
    period,
    semicolon,
    minus,
    minus_minus,
    minus_equals,
    plus,
    plus_plus,
    plus_equals,
    equals,
    equals_equals,
    less_than,
    less_than_equals,
    greater_than,
    greater_than_equals,
    keyword_var,
    keyword_if,
};

const keyword_lookup = std.ComptimeStringMap(TokenTag, .{
    .{ "var", TokenTag.keyword_var },
    .{ "if", TokenTag.keyword_if },
});

pub const Token = struct {
    tag: TokenTag,
    start: usize,
    end: usize,
    line: usize,
    col: usize,
};

pub const LexerError = error{
    UnexpectedCharacter,
} || std.mem.Allocator.Error;

pub const Lexer = struct {
    source: []const u8,
    index: usize = 0,
    line: usize = 1,
    col: usize = 1,
    queue: std.TailQueue(Token) = .{},
    allocator: *std.mem.Allocator,

    pub fn init(allocator: *std.mem.Allocator, source: []const u8) Lexer {
        return Lexer{
            .source = source,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *Lexer) void {
        while (self.queue.popFirst()) |node| {
            self.allocator.destroy(node);
        }
    }

    pub fn peekToken(self: *Lexer) ?*Token {
        if (self.queue.first) |node| {
            return &node.data;
        }
        return null;
    }

    pub fn nextToken(self: *Lexer) ?Token {
        if (self.queue.popFirst()) |node| {
            const token = node.data;
            self.allocator.destroy(node);
            return token;
        }
        return null;
    }

    pub fn tokenize(self: *Lexer) LexerError!void {
        while (self.index < self.source.len) {
            self.skipWhitespace();
            if (self.peekChar() == null) {
                break;
            }
            try self.tokenizeNext();
        }
    }

    fn tokenizeNext(self: *Lexer) LexerError!void {
        std.debug.assert(self.peekChar() != null);
        const next = self.peekChar().?;
        if (isNumber(next)) {
            try self.tokenizeNumber();
        } else if (isIdentifier(next)) {
            try self.tokenizeIdentifier();
        } else {
            try self.tokenizeSpecial();
        }
    }

    fn tokenizeNumber(self: *Lexer) LexerError!void {
        std.debug.assert(self.peekChar() != null);
        std.debug.assert(isNumber(self.peekChar().?));
        const start = self.index;
        while (self.peekChar()) |c| {
            if (!isNumber(c)) {
                break;
            }
            _ = self.nextChar();
        }
        const end = self.index;
        try self.pushToken(Token{ .tag = .number, .start = start, .end = end, .line = self.line, .col = self.col });
    }

    fn tokenizeIdentifier(self: *Lexer) LexerError!void {
        std.debug.assert(self.peekChar() != null);
        std.debug.assert(isIdentifier(self.peekChar().?));
        const start = self.index;
        while (self.peekChar()) |c| {
            if (!isIdentifier(c)) {
                break;
            }
            _ = self.nextChar();
        }
        const end = self.index;
        const tag = if (keyword_lookup.get(self.source[start..end])) |keyword| keyword else TokenTag.identifier;
        try self.pushToken(Token{ .tag = tag, .start = start, .end = end, .line = self.line, .col = self.col });
    }

    fn tokenizeSpecial(self: *Lexer) LexerError!void {
        std.debug.assert(self.peekChar() != null);
        const start = self.index;
        const char = self.nextChar().?;
        var tag: TokenTag = undefined;
        switch (char) {
            '(' => tag = .l_paren,
            ')' => tag = .r_paren,
            '{' => tag = .l_curly,
            '}' => tag = .r_curly,
            ',' => tag = .comma,
            '.' => tag = .period,
            '+' => if (self.peekChar()) |c| {
                switch (c) {
                    '+' => {
                        _ = self.nextChar();
                        tag = .plus_plus;
                    },
                    '=' => {
                        _ = self.nextChar();
                        tag = .plus_equals;
                    },
                    else => tag = .plus,
                }
            },
            '-' => if (self.peekChar()) |c| {
                switch (c) {
                    '-' => {
                        _ = self.nextChar();
                        tag = .minus_minus;
                    },
                    '=' => {
                        _ = self.nextChar();
                        tag = .minus_equals;
                    },
                    else => tag = .minus,
                }
            },
            '=' => if (self.peekChar()) |c| {
                switch (c) {
                    '=' => {
                        _ = self.nextChar();
                        tag = .equals_equals;
                    },
                    else => tag = .equals,
                }
            },
            '<' => if (self.peekChar()) |c| {
                switch (c) {
                    '=' => {
                        _ = self.nextChar();
                        tag = .less_than_equals;
                    },
                    else => tag = .less_than,
                }
            },
            '>' => if (self.peekChar()) |c| {
                switch (c) {
                    '=' => {
                        _ = self.nextChar();
                        tag = .greater_than_equals;
                    },
                    else => tag = .greater_than,
                }
            },
            ';' => tag = .semicolon,
            else => return LexerError.UnexpectedCharacter,
        }
        const end = self.index;
        try self.pushToken(Token{ .tag = tag, .start = start, .end = end, .line = self.line, .col = self.col });
    }

    fn pushToken(self: *Lexer, token: Token) LexerError!void {
        const node = try self.allocator.create(std.TailQueue(Token).Node);
        node.* = .{ .prev = null, .next = null, .data = token };
        self.queue.append(node);
    }

    fn peekChar(self: *Lexer) ?u8 {
        if (self.index >= self.source.len) {
            return null;
        }
        return self.source[self.index];
    }

    fn nextChar(self: *Lexer) ?u8 {
        if (self.index >= self.source.len) {
            return null;
        }
        const c = self.source[self.index];
        self.index += 1;
        self.col += 1;
        if (c == '\n') {
            self.line += 1;
            self.col = 1;
        }
        return c;
    }

    fn skipWhitespace(self: *Lexer) void {
        while (self.peekChar()) |c| {
            if (!isWhitespace(c)) {
                break;
            }
            _ = self.nextChar();
        }
    }

    fn isWhitespace(c: u8) bool {
        return switch (c) {
            '\n', '\t', '\r', ' ' => true,
            else => false,
        };
    }

    fn isIdentifier(c: u8) bool {
        return switch (c) {
            'a'...'z', 'A'...'Z', '0'...'9', '_' => true,
            else => false,
        };
    }

    fn isNumber(c: u8) bool {
        return switch (c) {
            '0'...'9' => true,
            else => false,
        };
    }
};
